multidownloader/types/set_sync_segment.go

```diff
@@ -0,0 +1,289 @@
+	for _, addr := range logQuery.Addrs {
+		segment := SyncSegment{
+			ContractAddr: addr,
+			BlockRange:   logQuery.BlockRange,
+		}
+		set.Add(segment)
+	}
+	return set
+}
+
+// Segments returns all SyncSegments in the SetSyncSegment
+func (s *SetSyncSegment) Segments() []SyncSegment {
+	result := make([]SyncSegment, 0, len(s.segments))
+	for _, segment := range s.segments {
+		result = append(result, *segment)
+	}
+	return result
+}
+
+// Add adds a new SyncSegment to the SetSyncSegment, merging block ranges
+// if the contract address already exists
+func (s *SetSyncSegment) Add(segment SyncSegment) {
+	// Check if exists
+	current := s.GetByContract(segment.ContractAddr)
+	if current == nil {
+		// Add new segment
+		s.segments = append(s.segments, &segment)
+		return
+	}
+	// Merge syncers
+	s.UpdateBlockRange(current, current.BlockRange.Merge(segment.BlockRange))
+}
+
+// GetByContract returns the SyncSegment for the given contract address
+func (s *SetSyncSegment) GetByContract(addr common.Address) *SyncSegment {
+	if s == nil {
+		return nil
+	}
+	for _, segment := range s.segments {
+		if segment.ContractAddr == addr {
+			return segment.Clone()
+		}
+	}
+	return nil
+}
+
+// SubtractSegments removes the block ranges defined in segments from the current SetSyncSegment
+// This is the pending data to synchronize
+func (f *SetSyncSegment) SubtractSegments(segments *SetSyncSegment) error {
+	if f == nil || segments == nil {
+		return nil
+	}
+	newSegments := f.Clone()
+	for _, segment := range segments.Segments() {
+		previousSegment := newSegments.GetByContract(segment.ContractAddr)
+		if previousSegment != nil {
+			brs := previousSegment.BlockRange.Subtract(segment.BlockRange)
+			switch len(brs) {
+			case 0:
+				newSegments.Remove(previousSegment)
+			case 1:
+				newSegments.UpdateBlockRange(previousSegment, brs[0])
+			default:
+				return fmt.Errorf("setSyncSegment.SubtractSegments: cannot split segment for %s into multiple ranges  %+v",
+					segment.String(), brs)
+			}
+		}
+	}
+	f.segments = newSegments.segments
+	return nil
+}
+
+// SubtractLogQuery removes the block ranges defined in the logQuery from the current SetSyncSegment
+// This is used to update the pendingSync after doing a FilterLogs query
+func (f *SetSyncSegment) SubtractLogQuery(logQuery *LogQuery) error {
+	if logQuery == nil {
+		return nil
+	}
+	newSegments := NewSetSyncSegmentFromLogQuery(logQuery)
+	return f.SubtractSegments(&newSegments)
+}
+
+// TotalBlocks returns the total number pending to synchronize
+func (f *SetSyncSegment) TotalBlocks() uint64 {
+	if f == nil {
+		return 0
+	}
+	minToBLock := ^uint64(0)
+	maxFromBlock := uint64(0)
+	for _, segment := range f.segments {
+		if segment.BlockRange.FromBlock < minToBLock {
+			minToBLock = segment.BlockRange.FromBlock
+		}
+		if segment.BlockRange.ToBlock > maxFromBlock {
+			maxFromBlock = segment.BlockRange.ToBlock
+		}
+	}
+	bn := aggkitcommon.NewBlockRange(minToBLock, maxFromBlock)
```

Comment:
Hmm, looking at this function it is not doing what its description is saying. It says it returns the total number of pending blocks to synchronize, but in fact, it returns a new block range spanning from the segment with lowest `FromBlock` to the highest `ToBlock`.

So for example if we have segments:
- `Contract A : 1-10 (10 blocks)`
- `Contract B: 100-110 (10 blocks)`
`TotalBlocks()` will return `1-110 (110)` blocks, even though it is syncing 20 blocks actually.

Maybe it is supposed to return the entire block range `1-110`, but the name and descriptions are misleading.